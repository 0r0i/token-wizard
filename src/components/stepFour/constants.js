export const MINTED_PREFIX = 'MINTED_CAPPED_'
export const DUTCH_PREFIX = 'DUTCH_'
export const DOWNLOAD_NAME = 'tokenwizard'
export const DOWNLOAD_TYPE = {
  text: 'text/plain',
  blob: 'blob'
}

export const ADDR_BOX_LEN = 44

export const TX_STEP_DESCRIPTION = {
  deployProxy: "Deploy proxy contract",
  crowdsaleCreate: "Create Crowdsale instance",
  token: "Initialize Token",
  setReservedTokens: "Register addresses for Reserved Tokens",
  updateGlobalMinContribution: "Update global minimum contribution",
  createCrowdsaleTiers: "Add tiers to Crowdsale",
  whitelist: "Register whitelisted addresses",
  crowdsaleInit: "Initialize Crowdsale",
}