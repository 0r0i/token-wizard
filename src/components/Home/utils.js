import { DOWNLOAD_STATUS, TOAST } from '../../utils/constants'
import { getNetworkVersion } from '../../utils/blockchainHelpers'
import { getCrowdsaleAssets } from '../../stores/utils'
import { toast } from '../../utils/utils'
import logdown from 'logdown'

const logger = logdown('TW:home:utils')

export const reloadStorage = async props => {
  let { generalStore, contractStore } = props

  try {
    contractStore.setProperty('downloadStatus', DOWNLOAD_STATUS.PENDING)

    // General store, check network
    let networkID = await getNetworkVersion()
    generalStore.setProperty('networkID', networkID)

    // Contract store, get contract and abi
    await getCrowdsaleAssets(networkID)
    contractStore.setProperty('downloadStatus', DOWNLOAD_STATUS.SUCCESS)
  } catch (e) {
    logger.error('Error downloading contracts', e)
    contractStore.setProperty('downloadStatus', DOWNLOAD_STATUS.FAILURE)
    throw e
  }
}
