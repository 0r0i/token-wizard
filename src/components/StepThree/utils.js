import moment from 'moment'
import { StepThreeFormMintedCapped } from './StepThreeFormMintedCapped'
import { StepThreeFormDutchAuction } from './StepThreeFormDutchAuction'
import { CROWDSALE_STRATEGIES } from '../../utils/constants'
import createDecorator from 'final-form-calculate'

export function defaultCompanyStartDate() {
  return moment()
    .add(5, 'minutes')
    .format('YYYY-MM-DDTHH:mm')
}

export const defaultCompanyEndDate = startDate => {
  return moment(startDate)
    .add(4, 'days')
    .startOf('day')
    .format('YYYY-MM-DDTHH:mm')
}

export const getStep3Component = strategy => {
  switch (strategy) {
    case CROWDSALE_STRATEGIES.MINTED_CAPPED_CROWDSALE:
      return StepThreeFormMintedCapped
    case CROWDSALE_STRATEGIES.DUTCH_AUCTION:
      return StepThreeFormDutchAuction
    default:
      return StepThreeFormMintedCapped
  }
}

export const tierDurationUpdater = tiers => {
  return createDecorator({
    field: /.+\.endTime/,
    updates: (value, name) => {
      const nextTierIndex = +name.match(/(\d+)/)[1] + 1
      const newValue = {}

      if (tiers[nextTierIndex]) {
        newValue[`tiers[${nextTierIndex}].startTime`] = value
      }

      return newValue
    }
  })
}
