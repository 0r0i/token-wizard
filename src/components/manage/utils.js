import { getCurrentAccount, sendTXToContract, methodToExec } from '../../utils/blockchainHelpers'
import {
  contractStore,
  crowdsaleStore,
  generalStore,
  tierStore,
  tokenStore,
  web3Store,
  reservedTokenStore
} from '../../stores'
import { VALIDATION_TYPES } from '../../utils/constants'
import { removeTrailingNUL, toFixed } from '../../utils/utils'
import { toBigNumber } from '../crowdsale/utils'
import moment from 'moment'

const { VALID } = VALIDATION_TYPES

const formatDate = timestamp => {
  return moment(timestamp * 1000).format('YYYY-MM-DDTHH:mm')
}

export const updateTierAttribute = async (attribute, value, tierIndex) => {
  let methodInterface
  let getParams
  const { decimals } = tokenStore
  const { isMintedCappedCrowdsale, isDutchAuction } = crowdsaleStore
  const methods = {
    startTime: isDutchAuction ? 'setCrowdsaleStartAndDuration' : null, // startTime is not changed after migration to Auth-os in MintedCappedCrowdsale strategy
    endTime: isMintedCappedCrowdsale ? 'updateTierDuration' : isDutchAuction ? 'setCrowdsaleStartAndDuration' : null,
    whitelist: isMintedCappedCrowdsale ? 'whitelistMultiForTier' : isDutchAuction ? 'whitelistMulti' : null,
    minCap: isMintedCappedCrowdsale ? 'updateTierMinimum' : isDutchAuction ? 'updateGlobalMinContribution' : null
  }

  let crowdsaleStartTime
  if (
    attribute === 'startTime' ||
    attribute === 'endTime' ||
    attribute === 'supply' ||
    attribute === 'whitelist' ||
    attribute === 'minCap'
  ) {
    if (attribute === 'startTime') {
      let { startTime, endTime } = tierStore.tiers[tierIndex]
      crowdsaleStartTime = toFixed(parseInt(Date.parse(value) / 1000, 10).toString())
      const duration = new Date(endTime) - new Date(startTime)
      const durationBN = (toBigNumber(duration) / 1000).toFixed()
      methodInterface = ['uint256', 'uint256']
      value = durationBN
      getParams = updateDutchAuctionDurationParams
    } else if (attribute === 'endTime') {
      let { startTime, endTime } = tierStore.tiers[tierIndex]
      console.log(startTime, endTime)
      const duration = new Date(endTime) - new Date(startTime)
      const durationBN = toBigNumber(duration).div(1000)
      value = durationBN.toFixed()
      methodInterface = ['uint256', 'uint256']
      if (isMintedCappedCrowdsale) {
        getParams = updateMintedCappedCrowdsaleDurationParams
      } else if (isDutchAuction) {
        getParams = updateDutchAuctionDurationParams
        crowdsaleStartTime = toFixed(new Date(startTime).getTime() / 1000).toString()
      }
    } else if (attribute === 'whitelist') {
      // whitelist
      const rate = tierStore.tiers[tierIndex].rate
      const rateBN = toBigNumber(rate)
      const oneTokenInETH = rateBN.pow(-1).toFixed()
      const oneTokenInWEI = web3Store.web3.utils.toWei(oneTokenInETH, 'ether')
      value = value.reduce(
        (toAdd, whitelist) => {
          toAdd[0].push(whitelist.addr)
          toAdd[1].push(
            toBigNumber(whitelist.min)
              .times(`1e${decimals}`)
              .toFixed()
          )
          toAdd[2].push(
            toBigNumber(whitelist.max)
              .times(`1e${decimals}`)
              .toFixed()
          )
          return toAdd
        },
        [[], [], []]
      )
      if (isMintedCappedCrowdsale) {
        methodInterface = ['uint256', 'address[]', 'uint256[]', 'uint256[]']
        getParams = updateTierWhitelistParams
      } else if (isDutchAuction) {
        methodInterface = ['address[]', 'uint256[]', 'uint256[]']
        getParams = updateWhitelistParams
      }
    } else if (attribute === 'minCap') {
      value = toBigNumber(tierStore.tiers[tierIndex].minCap)
        .times(`1e${tokenStore.decimals}`)
        .toFixed()

      if (isMintedCappedCrowdsale) {
        methodInterface = ['uint256', 'uint256']
        getParams = updateTierMinimumParams
      } else if (isDutchAuction) {
        methodInterface = ['uint256']
        getParams = updateMinimumParams
      }
    }
  }

  console.log('crowdsaleStartTime:', crowdsaleStartTime)
  console.log('value:', value)

  console.log('attribute:', attribute)
  console.log('methods[attribute]:', methods[attribute])

  console.log('tierIndex:', tierIndex)

  let paramsToExec
  if (isMintedCappedCrowdsale) {
    paramsToExec = [tierIndex, value, methodInterface]
  } else if (isDutchAuction) {
    if (attribute === 'endTime') {
      paramsToExec = [crowdsaleStartTime, value, methodInterface]
    } else {
      paramsToExec = [value, methodInterface]
    }
  }

  console.log('paramsToExec:', paramsToExec)
  console.log('methods[attribute]:', methods[attribute])
  console.log('methodInterface:', methodInterface)

  let targetContractName
  if (crowdsaleStore.execID) {
    targetContractName = 'registryExec'
  } else {
    targetContractName = 'MintedCappedProxy'
  }

  const method = methodToExec(
    targetContractName,
    `${methods[attribute]}(${methodInterface.join(',')})`,
    getParams,
    paramsToExec
  )
  console.log('method:', method)

  const account = await getCurrentAccount()
  const opts = { gasPrice: generalStore.gasPrice, from: account }
  opts.gasLimit = await method.estimateGas(opts)

  return sendTXToContract(method.send(opts))
}

const updateMintedCappedCrowdsaleDurationParams = (tierIndex, duration, methodInterface) => {
  console.log(tierIndex, duration)
  return web3Store.web3.eth.abi.encodeParameters(methodInterface, [tierIndex, duration])
}

const updateDutchAuctionDurationParams = (startTime, duration, methodInterface) => {
  console.log(startTime, duration)
  return web3Store.web3.eth.abi.encodeParameters(methodInterface, [startTime, duration])
}

const updateTierWhitelistParams = (tierIndex, [addr, min, max], methodInterface) => {
  console.log(tierIndex, addr, min, max, methodInterface)
  return web3Store.web3.eth.abi.encodeParameters(methodInterface, [tierIndex, addr, min, max])
}

const updateWhitelistParams = ([addr, min, max], methodInterface) => {
  console.log(addr, min, max, methodInterface)
  return web3Store.web3.eth.abi.encodeParameters(methodInterface, [addr, min, max])
}

const updateTierMinimumParams = (tierIndex, minCap, methodInterface) => {
  console.log(tierIndex, minCap, methodInterface)
  return web3Store.web3.eth.abi.encodeParameters(methodInterface, [tierIndex, minCap])
}

const updateMinimumParams = (minCap, methodInterface) => {
  console.log(minCap, methodInterface)
  return web3Store.web3.eth.abi.encodeParameters(methodInterface, [minCap])
}

const crowdsaleData = (tier, crowdsale, token, reserved_tokens_info) => {
  const { isMintedCappedCrowdsale } = crowdsaleStore
  const { toAscii } = web3Store.web3.utils
  const {
    start_time,
    tier_start,
    end_time,
    tier_end,
    current_rate,
    tier_price,
    tier_sell_cap,
    tier_min,
    tokens_remaining,
    tokens_sold,
    whitelist,
    duration_is_modifiable,
    is_whitelisted
  } = tier
  let { tier_name } = tier

  try {
    tier_name = removeTrailingNUL(toAscii(tier_name))
  } catch (e) {
    console.log('###Token name is already in ASCII###')
  }

  const { team_wallet, is_finalized } = crowdsale
  const { total_supply, token_decimals } = token
  let { token_name, token_symbol } = token

  try {
    token_name = removeTrailingNUL(toAscii(token_name))
  } catch (e) {
    console.log('###Token name is already in ASCII###')
  }

  try {
    token_symbol = removeTrailingNUL(toAscii(token_symbol))
  } catch (e) {
    console.log('###Token name is already in ASCII###')
  }

  return {
    wallet: team_wallet,
    start_time: isMintedCappedCrowdsale ? tier_start : start_time,
    end_time: isMintedCappedCrowdsale ? tier_end : end_time,
    rate: isMintedCappedCrowdsale ? tier_price : current_rate,
    max_sell_cap: isMintedCappedCrowdsale
      ? tier_sell_cap
      : toBigNumber(tokens_remaining)
          .plus(tokens_sold || '0')
          .toFixed(),
    min_cap: toBigNumber(tier_min)
      .div(`1e${token_decimals}`)
      .toFixed(),
    name: isMintedCappedCrowdsale ? tier_name : '',
    updatable: isMintedCappedCrowdsale ? duration_is_modifiable : true,
    whitelist: whitelist || [],
    is_whitelisted,
    finalized: is_finalized,
    crowdsale_token: {
      name: token_name,
      ticker: token_symbol,
      supply: toBigNumber(total_supply)
        .div(`1e${token_decimals}`)
        .toFixed(),
      decimals: token_decimals,
      reserved_accounts: reserved_tokens_info
    }
  }
}

export const processTier = (crowdsale, token, reserved_tokens_info, tier, tier_index) => {
  console.log('tier:', tier)
  console.log('reserved_tokens_info:', reserved_tokens_info)
  console.log('crowdsale:', crowdsale)
  console.log('token:', token)

  const { web3 } = web3Store
  const _crowdsaleData = crowdsaleData(tier, crowdsale, token, reserved_tokens_info)
  const {
    wallet,
    start_time,
    end_time,
    rate: rate_in_wei,
    max_sell_cap,
    min_cap,
    name,
    whitelist,
    updatable,
    is_whitelisted,
    finalized,
    crowdsale_token
  } = _crowdsaleData
  console.log(_crowdsaleData)
  console.log('reserved_tokens_info:', crowdsale_token.reserved_accounts)

  const token_decimals = !isNaN(crowdsale_token.decimals) ? crowdsale_token.decimals : 0
  const max_cap_before_decimals = toBigNumber(max_sell_cap)
    .div(`1e${token_decimals}`)
    .toFixed()
  const rate =
    rate_in_wei > 0
      ? toBigNumber(web3.utils.fromWei(rate_in_wei, 'ether'))
          .pow(-1)
          .dp(0)
          .toFixed()
      : 0
  // TODO: remove this filter after Auth-os implement uniqueness for the whitelisted addresses (#871)
  const filtered_whitelist = [...new Set(whitelist.map(item => JSON.stringify(item)))].map(item => JSON.parse(item))

  const new_tier = {
    tier: name,
    walletAddress: wallet,
    rate,
    supply: max_cap_before_decimals || 0,
    minCap: min_cap,
    startTime: formatDate(start_time),
    endTime: formatDate(end_time),
    updatable,
    whitelistEnabled: is_whitelisted ? 'yes' : 'no',
    whitelist: filtered_whitelist.map(({ addr, min, max }) => ({
      addr,
      min: toBigNumber(min)
        .div(`1e${token_decimals}`)
        .toFixed(),
      max: toBigNumber(max)
        .div(`1e${token_decimals}`)
        .toFixed(),
      stored: true
    }))
  }

  const validations = {
    tier: VALID,
    walletAddress: VALID,
    rate: VALID,
    supply: VALID,
    startTime: VALID,
    endTime: VALID,
    updatable: VALID
  }

  const initial_tier_values = {
    duration: end_time * 1000 - start_time * 1000,
    updatable,
    index: tier_index,
    startTime: new_tier.startTime,
    endTime: new_tier.endTime,
    whitelist: new_tier.whitelist.slice(),
    isWhitelisted: is_whitelisted,
    supply: new_tier.supply,
    minCap: new_tier.minCap,
    addresses: {
      crowdsaleAddress: contractStore.crowdsale.execID
    }
  }

  tierStore.addTier(new_tier, validations)
  tierStore.sortWhitelist(tier_index)

  crowdsaleStore.addInitialTierValues(initial_tier_values)
  crowdsaleStore.setSelectedProperty('finalized', finalized)
  crowdsaleStore.setSelectedProperty('updatable', crowdsaleStore.selected.updatable || updatable)

  crowdsale_token.reserved_accounts.forEach(reservedTokenStore.addToken)

  tokenStore.setProperty('name', crowdsale_token.name)
  tokenStore.setProperty('ticker', crowdsale_token.ticker)
  tokenStore.setProperty('decimals', crowdsale_token.decimals)
  tokenStore.setProperty('supply', crowdsale_token.supply)
}

export function getFieldsToUpdate(updatableTiers, tiers) {
  if (!updatableTiers.length || !tiers.length) return []

  const keys = Object.keys(updatableTiers[0]).filter(
    key => key === 'endTime' || key === 'whitelist' || key === 'minCap'
  )

  return updatableTiers.reduce((toUpdate, updatableTier, index) => {
    keys.forEach(key => {
      let newValue = tiers[updatableTier.index][key]

      if (key === 'whitelist') {
        newValue = newValue.filter(item => !item.stored)

        if (newValue.length) {
          toUpdate.push({ key, newValue, tier: index })
        }
      } else if (key === 'endTime') {
        const end = new Date(tiers[updatableTier.index].endTime)
        const start = new Date(tiers[updatableTier.index].startTime)
        const duration = end - start

        if (updatableTier.duration !== duration) {
          toUpdate.push({ key, newValue, tier: index })
        }
      } else if (key === 'minCap') {
        if (newValue !== updatableTier.minCap) {
          toUpdate.push({ key, newValue, tier: index })
        }
      }
    })

    return toUpdate
  }, [])
}
