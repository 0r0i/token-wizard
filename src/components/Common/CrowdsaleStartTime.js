import React from 'react'
import { Field } from 'react-final-form'
import { InputField2 } from './InputField2'
import { validateTierStartDate } from '../../utils/validations'
import { DESCRIPTION, TEXT_FIELDS } from '../../utils/constants'

export const CrowdsaleStartTime = ({ index, ...props }) => (
  <Field
    component={InputField2}
    description={props.description || DESCRIPTION.START_TIME}
    label={TEXT_FIELDS.START_TIME}
    type="datetime-local"
    validate={props.disabled ? undefined : validateTierStartDate(index)}
    {...props}
  />
)
