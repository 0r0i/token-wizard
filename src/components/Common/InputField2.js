import React from 'react'
import { Errors } from './Errors'
import { FormControlTitle } from '../Common/FormControlTitle'
import { TextField } from '../Common/TextField'

export const InputField2 = props => {
  if (props.val) {
    props.input.value = props.val
  }

  return (
    <div className={`sw-InputField2 ${props.extraClassName ? props.extraClassName : ''}`}>
      {props.label ? <FormControlTitle title={props.label} description={props.description} /> : null}
      <TextField
        autoComplete="off"
        className={props.inputClassName ? props.inputClassName : 'input'}
        disabled={props.disabled}
        id={props.input.name}
        placeholder={props.placeholder}
        type={props.type}
        min={props.min}
        max={props.max}
        {...props.input}
      />
      <Errors name={props.input.name} />
    </div>
  )
}
