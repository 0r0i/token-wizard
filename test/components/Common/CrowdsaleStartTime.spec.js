import React from 'react'
import { CrowdsaleStartTime } from '../../../src/components/Common/CrowdsaleStartTime'
import { Form } from 'react-final-form'
import Adapter from 'enzyme-adapter-react-15'
import { configure, mount, shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import { VALIDATION_MESSAGES, TEXT_FIELDS, DESCRIPTION } from '../../../src/utils/constants';
import MockDate from 'mockdate'
import { InputField2 } from '../../../src/components/Common/InputField2'

configure({ adapter: new Adapter() })
const DECRIPTION = DESCRIPTION.START_TIME
const LABEL = TEXT_FIELDS.START_TIME

describe('CrowdsaleStartTime ', () => {
  const crowdsale = { tiers: [{ endTime: Date().now }] }
  it(`should render CrowdsaleStartTime component`, () => {
    const wrapper = renderer.create(
      <Form
        onSubmit={jest.fn()}
        component={CrowdsaleStartTime}
        disabled={false}
        initialValues={crowdsale}
        errorStyle={{ color: 'red', fontWeight: 'bold' }}
        name="name"
        index={0}
        disabled={false}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it(`should render MinCap component if field is disabled`, () => {
    const wrapper = renderer.create(
      <Form
        onSubmit={jest.fn()}
        component={CrowdsaleStartTime}
        disabled={true}
        initialValues={crowdsale}
        errorStyle={{ color: 'red', fontWeight: 'bold' }}
        name="name"
        index={0}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it(`should be  proper label and description`, () => {
    const wrapper = mount(
      <Form
        onSubmit={jest.fn()}
        component={CrowdsaleStartTime}
        initialValues={crowdsale}
        errorStyle={{ color: 'red', fontWeight: 'bold' }}
        name="name"
        disabled={false}
        index={0}
      />
    )
    let label = wrapper.find('label')
    expect(label.text()).toBe(LABEL)
    let descript = wrapper.find('p[className="description"]')
    expect(descript.text()).toBe(DECRIPTION)
  })
})

describe('CrowdsaleStartTime ', () => {
  const TIMESTAMPS = {
    CURRENT_TIME: 1520852400000,
    PLUS_5_MINUTES: 1520852700000,
    PLUS_10_MINUTES: 1520853000000,
    PLUS_10_DAYS: 1521716400000,
    MINUS_5_MINUTES: 1520852100000,
    MINUS_10_DAYS: 1519988400000
  }
  beforeEach(() => {
    MockDate.set(TIMESTAMPS.CURRENT_TIME)
  })
  afterEach(() => {
    MockDate.reset()
  })
  const crowdsale = { tiers: [{ endTime: TIMESTAMPS.PLUS_10_MINUTES }] }
  const wrapper = mount(
    <Form
      onSubmit={jest.fn()}
      initialValues={crowdsale}
      component={CrowdsaleStartTime}
      errorStyle={{ color: 'red', fontWeight: 'bold' }}
      name="startTime"
      index={0}
      disabled={false}
    />
  )
  it(`shouldn't be errors if value is correct`, () => {
    const input = wrapper.find('input[name="startTime"]')
    input.simulate('change', { target: { value: TIMESTAMPS.PLUS_5_MINUTES } })
    const inputProps = wrapper.find('InputField2').props()
    expect(inputProps.meta.error).toBeUndefined()
  })
  it(`should fail if value is empty`, () => {
    const input = wrapper.find('input[name="startTime"]')
    input.simulate('change', { target: { value: '' } })
    const inputProps = wrapper.find('InputField2').props()
    expect(inputProps.meta.error[0]).toBe(VALIDATION_MESSAGES.REQUIRED)
  })
  it(`should fail if startTime is previous than current time`, () => {
    const input = wrapper.find('input[name="startTime"]')
    input.simulate('change', { target: { value: TIMESTAMPS.MINUS_5_MINUTES } })
    const inputProps = wrapper.find('InputField2').props()
    expect(inputProps.meta.error[0]).toBe(VALIDATION_MESSAGES.DATE_IN_FUTURE)
  })

  it(`should fail if startTime is  later than end time`, () => {
    const input = wrapper.find('input[name="startTime"]')
    input.simulate('change', { target: { value: TIMESTAMPS.PLUS_10_DAYS } })
    const inputProps = wrapper.find('InputField2').props()
    expect(inputProps.meta.error[0]).toBe("Should be previous than same tier's End Time")
  })
  it(`should fail if startTime is same with end time`, () => {
    const input = wrapper.find('input[name="startTime"]')
    input.simulate('change', { target: { value: TIMESTAMPS.PLUS_10_MINUTES } })
    const inputProps = wrapper.find('InputField2').props()
    expect(inputProps.meta.error[0]).toBe("Should be previous than same tier's End Time")
  })

  it(`should fail if startTime is same or later than previous tier's endTime`, () => {
    const crowdsale = {
      tiers: [{ endTime: TIMESTAMPS.PLUS_10_MINUTES },
        { endTime: TIMESTAMPS.PLUS_10_DAYS }]
    }
    const wrapper = mount(
      <Form
        onSubmit={jest.fn()}
        initialValues={crowdsale}
        component={CrowdsaleStartTime}
        errorStyle={{ color: 'red', fontWeight: 'bold' }}
        name="startTime"
        index={1}
        disabled={false}
      />
    )
    const input = wrapper.find('input[name="startTime"]')
    input.simulate('change', { target: { value: TIMESTAMPS.PLUS_5_MINUTES } })
    const inputProps = wrapper.find('InputField2').props()
    expect(inputProps.meta.error[0]).toBe("Should be same or later than previous tier's End Time")
  })
})
