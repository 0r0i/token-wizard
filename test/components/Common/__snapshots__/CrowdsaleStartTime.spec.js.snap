// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CrowdsaleStartTime  should render CrowdsaleStartTime component 1`] = `
<div
  className={undefined}
>
  <label
    className="label"
    htmlFor="name"
  >
    Start Time
  </label>
  <input
    autoComplete="off"
    className="input"
    disabled={false}
    id="name"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder={undefined}
    type="datetime-local"
    value=""
  />
  <p
    className="description"
  >
    Date and time when the tier starts. Can't be in the past from the current moment.
  </p>
  <span>
    <p
      className="error"
      style={
        Object {
          "color": "red",
          "fontWeight": "bold",
        }
      }
    />
    <p
      className="error"
      style={
        Object {
          "color": "red",
          "fontWeight": "bold",
        }
      }
    />
    <p
      className="error"
      style={
        Object {
          "color": "red",
          "fontWeight": "bold",
        }
      }
    />
  </span>
</div>
`;

exports[`CrowdsaleStartTime  should render MinCap component if field is disabled 1`] = `
<div
  className={undefined}
>
  <label
    className="label"
    htmlFor="name"
  >
    Start Time
  </label>
  <input
    autoComplete="off"
    className="input"
    disabled={true}
    id="name"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder={undefined}
    type="datetime-local"
    value=""
  />
  <p
    className="description"
  >
    Date and time when the tier starts. Can't be in the past from the current moment.
  </p>
  <span>
    <p
      className="error"
      style={
        Object {
          "color": "red",
          "fontWeight": "bold",
        }
      }
    />
  </span>
</div>
`;
